front50-api/src/main/java/com/netflix/spinnaker/front50/api/validator/PipelineValidator.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright 2021 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.front50.api.validator;
+
+import com.netflix.spinnaker.front50.api.model.pipeline.Pipeline;
+import com.netflix.spinnaker.kork.annotations.Alpha;
+import com.netflix.spinnaker.kork.plugins.api.internal.SpinnakerExtensionPoint;
+import com.netflix.spinnaker.kork.web.exceptions.ValidationException;
+
+/**
+ * A {@link PipelineValidator} provides a hook where custom validation can be applied to pipeline
+ * pre-save/update operations.
+ */
+@Alpha
+public interface PipelineValidator extends SpinnakerExtensionPoint {
+  /**
+   * @param pipeline the pipeline being created/modified
+   * @throws ValidationException to fail validation
+   */
+  void validate(Pipeline pipeline);
```

Comment:
Decided that the pipeline validators should throw a `ValidationException` so we don't need `org.springframework.validation.Errors` here.

Reply:
I am concerned that Netflix will not like this change. It doesn't look like there are any public implementations in Front50, which likely means that the real implementations live in Netflix's extension projects. You could open the PR and see what they think, but it's possible that they'll ask you to provide an `Errors` class that implements the same methods as Spring's `Errors` class (in the same way that you reimplemented `ForwardingMap`.

Reply:
Thanks for letting me know! I'd rather write something more friendly for Netflix so I'll look into implementing the Errors class.

Reply:
It looks like `NESTED_PATH_SEPARATOR` in `org.springframework.validation.Errors` imports `org.springframework.beans.PropertyAccessor`. Not sure if I can get around that other than leaving that out. Not sure what sort of internal implementations Netflix has, but hopefully none that are dependent on `NESTED_PATH_SEPARATOR`.